sugarloaf/src/components/rich_text/batch.rs

```diff
@@ -0,0 +1,394 @@
+use bytemuck::{Zeroable, Pod};
+use crate::components::rich_text::color::Color;
+use crate::components::rich_text::image_cache::TextureId;
+
+/// Batch geometry vertex.
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Zeroable, Pod)]
+pub struct Vertex {
+    pub pos: [f32; 4],
+    pub color: [f32; 4],
+    pub uv: [f32; 2],
+}
+
+/// Rectangle with floating point coordinates.
+#[derive(Copy, Clone, Default, Debug)]
+pub struct Rect {
+    pub x: f32,
+    pub y: f32,
+    pub width: f32,
+    pub height: f32,
+}
+
+impl Rect {
+    /// Creates a new rectangle.
+    pub fn new(x: f32, y: f32, width: f32, height: f32) -> Self {
+        Rect {
+            x,
+            y,
+            width,
+            height,
+        }
+    }
+}
+
+impl From<[f32; 4]> for Rect {
+    fn from(v: [f32; 4]) -> Self {
+        Self::new(v[0], v[1], v[2], v[3])
+    }
+}
+
+#[derive(Default, Debug)]
+struct Batch {
+    image: Option<TextureId>,
+    mask: Option<TextureId>,
+    vertices: Vec<Vertex>,
+    indices: Vec<u32>,
+    subpix: bool,
+}
+
+impl Batch {
+    fn clear(&mut self) {
+        self.image = None;
+        self.mask = None;
+        self.vertices.clear();
+        self.indices.clear();
+        self.subpix = false;
+    }
+
+    fn add_rect(
+        &mut self,
+        rect: &Rect,
+        depth: f32,
+        color: Color,
+        coords: Option<&[f32; 4]>,
+        image: Option<TextureId>,
+        mask: Option<TextureId>,
+        subpix: bool,
+    ) -> bool {
+        if !self.vertices.is_empty() && subpix != self.subpix {
+            return false;
+        }
+        let has_image = image.is_some();
+        let has_mask = mask.is_some();
+        if has_image {
+            if self.image.is_some() && self.image != image {
+                return false;
+            }
+        }
+        if has_mask {
+            if self.mask.is_some() && self.mask != mask {
+                return false;
+            }
+        }
+        self.subpix = subpix;
+        let flags = match (has_image, has_mask) {
+            (true, true) => {
+                self.image = image;
+                self.mask = mask;
+                3.
+            }
+            (true, false) => {
+                self.image = image;
+                1.
+            }
+            (false, true) => {
+                self.mask = mask;
+                2.
+            }
+            _ => 0.,
+        };
```

Comment:
Can't you save a few operations in the worst case scenarios by merging the `match` and `ifs`?

```rust
let mut flags = 0.0;

if has_image {
    if self.image.is_some() && self.image != image {
        return false;
    }

    self.image = image;

    flags += 1.0;
}

if has_mask {
    if self.mask.is_some() && self.mask != mask {
        return false;
    }

    self.mask = mask;

    flags += 2.0;
}

self.subpix = subpix;
```